Automattic/woocommerce-payments
# Detect multiple enabled payment methods of the same type

Closes https://github.com/Automattic/woocommerce-payments/issues/8384

#### Changes proposed in this Pull Request
This PR is the first iteration of the feature which should help merchants identify payment methods of the same type that are enabled by multiple plugins. It implements the simple solution mentioned in paJDYF-ck0-p2. The inline notice pops up under the payment method and can be dismissed. Once dismissed, it won't appear again upon page refresh. 

> Note: as mentioned in the solution description, the downside of the simple solution implemented in this PR is that we base on the hardcoded keywords list. Due to the fact that there are currently no rules for gateways (internal & third-party) to follow in terms of gateway ID, this solution cannot guarantee its effectiveness in every use case, as some third-party plugin is free to call card gateway in any other way, e.g. `third-party-gateway`. This limitation is known.

For testing steps outlined below, you can use multiple plugins. During development, I used WooPayments, [Mollie](https://wordpress.org/plugins/mollie-payments-for-woocommerce/), [Stripe](https://github.com/woocommerce/woocommerce-gateway-stripe) and [Klarna](https://wordpress.org/plugins/klarna-payments-for-woocommerce/)). If you'd like to quickly setup Mollie, contact @timur27 for test API keys which hopefully will save you a bit of time. For more inspiration on which plugins you could use, please see paJDYF-ck0-p2#comment-23134 which refers the duplicates data.









#### Testing instructions


```
💡 The inline notice text was confirmed with Product & Design and is following: 
This payment method is enabled by other extensions. Review extensions to improve the shopper experience.

Here's how the inline notice looks like:
```
<p align="center">
<img width="501" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/22319444/9fd098c6-4fcc-489b-8aab-67743fc813c5">
</p>

* Enable credit card by any other plugin e.g. Mollie or Stripe
* Navigate to Payments settings and confirm there's an inline notice under credit card payment method. _Review extensions_ should open the WooCommerce payment settings page
* Enable any other payment method from APMs list. You can use Klarna gateway to check with Klarna or Stripe for any other payment method.
* Confirm that the notice appears.
* Test the scenario when a payment method is enabled in three different places, including WooPayments. At this point, you will see the notice. However, if we disable payment method by WooPayments, the notice should disappear, although it's still enabled by two other plugins. The goal is to show the notice only if WooPayments is part of the plugins group that introduce a duplicate.
* Check Apple Pay/Google Pay by enabling them e.g. under Stripe gateway and checking if the notice appears. Stripe gateway allows PRBs to be enabled with both disabled and enabled CC, and both use-cases should be covered by this PR.
* Dismiss any notice and confirm that it's not shown after closing. Refresh the page and confirm no notice is present.
* Confirm that notice is handled separately, meaning that only part of the notices can be dismissed. After such action, the rest should still be present on the page.

-------------------


**Post merge**
